File: lib/Models/BranchModels.dart
import 'BaseResponse.dart';
import 'FoodModels.dart';

// ==================== BranchModel ====================
class BranchModel {
  BranchModel({
    required this.id,
    required this.name,
    required this.address,
    required this.phone,
    this.imageUrl,
    this.openTime,
    this.closeTime,
  });

  final String id;
  final String name;
  final String address;
  final String phone;
  final String? imageUrl;
  final String? openTime;
  final String? closeTime;

  factory BranchModel.fromJson(Map<String, dynamic> json) {
    return BranchModel(
      id: json['id'] as String,
      name: json['name'] as String,
      address: json['address'] as String,
      phone: json['phone'] as String,
      imageUrl: json['imageUrl'] as String?,
      openTime: json['openTime'] as String?,
      closeTime: json['closeTime'] as String?,
    );
  }
}

// ==================== BranchFoodModel ====================
class BranchFoodModel {
  BranchFoodModel({
    required this.id,
    required this.price,
    required this.active,
    required this.branch,
    required this.food,
  });

  final String id;
  final int price;
  final bool active;
  final BranchModel branch;
  final FoodModel food;

  factory BranchFoodModel.fromJson(Map<String, dynamic> json) {
    return BranchFoodModel(
      id: json['id'] as String,
      price: (json['price'] as num).toInt(),
      active: json['active'] as bool,
      branch: BranchModel.fromJson(json['branch'] as Map<String, dynamic>),
      food: FoodModel.fromJson(json['food'] as Map<String, dynamic>),
    );
  }
}

// ==================== BranchListResponse ====================
class BranchListResponse extends BaseResponse {
  late List<BranchModel> data;

  BranchListResponse({
    required int statusCode,
    required String message,
    List<BranchModel> data = const [],
  }) : super(statusCode: statusCode, message: message) {
    this.data = data;
  }

  factory BranchListResponse.fromJson(Map<String, dynamic> json) {
    final raw = json['data'];
    final list = raw is List ? raw : const [];
    return BranchListResponse(
      statusCode: (json['statusCode'] as num).toInt(),
      message: json['message'] as String,
      data: list
          .whereType<Map<String, dynamic>>()
          .map(BranchModel.fromJson)
          .toList(),
    );
  }
}

// ==================== BranchFoodPageResponse ====================
class BranchFoodPageResponse extends BaseResponse {
  late List<BranchFoodModel> data;
  PaginationMeta? meta;

  BranchFoodPageResponse({
    required int statusCode,
    required String message,
    List<BranchFoodModel> data = const [],
    this.meta,
  }) : super(statusCode: statusCode, message: message) {
    this.data = data;
  }

  factory BranchFoodPageResponse.fromJson(Map<String, dynamic> json) {
    final payload = json['data'];
    final dataMap = payload is Map<String, dynamic> ? payload : <String, dynamic>{};
    final result = dataMap['result'];
    final list = result is List ? result : const [];
    final metaRaw = dataMap['meta'];
    return BranchFoodPageResponse(
      statusCode: (json['statusCode'] as num).toInt(),
      message: json['message'] as String,
      meta: metaRaw is Map<String, dynamic> ? PaginationMeta.fromJson(metaRaw) : null,
      data: list
          .whereType<Map<String, dynamic>>()
          .map(BranchFoodModel.fromJson)
          .toList(),
    );
  }
}

File: mobile_restaurant/models/BranchModels.ts
import { buildApiBasePath } from '../utils/ApiConfig';
import { parseBaseResponse, type BaseResponse } from './AuthModels';
import { parseFoodModel, parsePaginationMeta, type FoodModel, type PaginationMeta } from './FoodModels';

function resolveBranchImageUrl(fileName?: string): string | undefined {
  if (!fileName || !fileName.trim()) return undefined;
  if (/^https?:\/\//i.test(fileName)) return fileName.trim();
  return `${buildApiBasePath()}/storage/branch/${fileName.trim()}`;
}

// ==================== BranchModel ====================
export interface BranchModel {
  id: string;
  name: string;
  address: string;
  phone: string;
  imageUrl?: string;
  openTime?: string;
  closeTime?: string;
  displayImageUrl?: string;
}

export function parseBranchModel(json: any): BranchModel {
  return {
    id: json.id,
    name: json.name,
    address: json.address,
    phone: json.phone,
    imageUrl: json.imageUrl ?? undefined,
    openTime: json.openTime ?? undefined,
    closeTime: json.closeTime ?? undefined,
    displayImageUrl: resolveBranchImageUrl(json.imageUrl),
  };
}

// ==================== BranchFoodModel ====================
export interface BranchFoodModel {
  id: string;
  price: number;
  active: boolean;
  branch: BranchModel;
  food: FoodModel;
}

export function parseBranchFoodModel(json: any): BranchFoodModel {
  return {
    id: json.id,
    price: Number(json.price),
    active: json.active,
    branch: parseBranchModel(json.branch),
    food: parseFoodModel(json.food),
  };
}

// ==================== BranchListResponse ====================
export interface BranchListResponse extends BaseResponse {
  data: BranchModel[];
}

export function parseBranchListResponse(json: any): BranchListResponse {
  const base = parseBaseResponse(json);
  const raw = json.data;
  const list = Array.isArray(raw) ? raw : [];
  return {
    ...base,
    data: list.map(parseBranchModel),
  };
}

// ==================== BranchFoodPageResponse ====================
export interface BranchFoodPageResponse extends BaseResponse {
  data: BranchFoodModel[];
  meta?: PaginationMeta;
}

export function parseBranchFoodPageResponse(json: any): BranchFoodPageResponse {
  const base = parseBaseResponse(json);
  const payload = json.data;
  const dataMap = payload && typeof payload === 'object' && !Array.isArray(payload) ? payload : {};
  const result = dataMap.result;
  const list = Array.isArray(result) ? result : [];
  return {
    ...base,
    data: list.map(parseBranchFoodModel),
    meta: dataMap.meta ? parsePaginationMeta(dataMap.meta) : undefined,
  };
}
